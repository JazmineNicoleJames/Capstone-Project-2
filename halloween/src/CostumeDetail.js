import React, {useEffect, useState} from "react";
import { useParams } from "react-router-dom";
import YouTubeVideo from "./YouTubeVideo";
import AmazonLink from "./AmazonLink";

/* CostumeDetail component 
    * uses allCostumes as a prop
        
    * uses the params (item_name) to filter through allCostumes - saves it as costumeData
    
    * uses costumeData to set selectedCostume 

    * displays Loading... while fetching data 

    * renders either YouTubeVideo or AmazonLink or both 
        -- depending on result_type from selectedCostume data

*/

const CostumeDetail = ({ allCostumes }) => {

    const {item_name} = useParams();
    const [selectedCostume, setSelectedCostume] = useState(null);
    const [error, setError] = useState(null);
    
    useEffect(() => {
        const fetchCostume = async () => {
            try{
                if(allCostumes !== undefined){
                    const costumeData = allCostumes.find(costume => costume.item_name === item_name);
                    setSelectedCostume(costumeData);
                }

            } catch (e) {
                setError(e)
                console.error('e', e)
            }

        };
        fetchCostume();
    }, [allCostumes, item_name]);

    if(error){
        return(<div> Error </div>)
    }
    
    if(!selectedCostume) return <div>Loading...</div>;

    const {result_type, youtube_video_id, amazon_link_id, img} = selectedCostume;
    return (
        <div>
            {result_type === 'youtube-video' && youtube_video_id && (
                <YouTubeVideo videoId={youtube_video_id} />
            )}
            {result_type === 'amazon-link' && amazon_link_id && (
                <AmazonLink amazonLink={amazon_link_id} image={img} />
            )}
            {result_type === 'both' && (
                <div>
                <YouTubeVideo videoId={youtube_video_id} />
                <AmazonLink amazonLink={amazon_link_id} />
                </div>
            )}
        </div>
    )
};

export default CostumeDetail;