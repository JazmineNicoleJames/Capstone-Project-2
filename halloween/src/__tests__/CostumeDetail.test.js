import { MemoryRouter, Route, Routes } from "react-router-dom";
import {render, waitFor } from '@testing-library/react';
import CostumeDetail from "../CostumeDetail";


jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
      item_name: 'fake'
  })
}));

jest.mock('../api', () => ({
  __esModule: true,
  default: {
      getCostume: jest.fn().mockResolvedValue(
        {
          item_name: 'fake',
          title: "testing title",
          description: "testing description",
          result_type: 'youtube-video',
          youtube_video_id: 'youtube-video-id'
        }
      ),
      getCostumes: jest.fn().mockResolvedValue(
        {
          item_name: 'fake'
        }
      )
  }
}));

test("should match snapshot", async () => {
  const allCostumes = [
    {
      item_name: 'fake',
      title: "testing title",
      description: "testing description",
      result_type: 'youtube-video',
      youtube_video_id: 'youtube-video-id'
    }
  ];

  const { asFragment } = render(
      <MemoryRouter>
          <CostumeDetail allCostumes={allCostumes}/>
      </MemoryRouter>
  );

  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  })
});


test("renders loading message while loading", () => {
  const { getByText } = render(
    <MemoryRouter>
      <CostumeDetail />
    </MemoryRouter>
  );
  expect(getByText("Loading...")).toBeInTheDocument();
});


test("renders single costume", async () => {
  const allCostumes = [
    {
      item_name: 'fake',
      title: "testing title",
      description: "testing description",
      result_type: 'youtube-video',
      youtube_video_id: 'youtube-video-id'
    }
  ];

  const { getByText } = render(
    <MemoryRouter initialEntries={['/costumes/fake']}>
      <Routes>
      <Route path="/costumes/:item_name" element={<CostumeDetail allCostumes={allCostumes} />}>
      </Route>
      </Routes>
    </MemoryRouter>
  );

  await waitFor(() => {
    expect(getByText("FAKE")).toBeInTheDocument()
  })
});