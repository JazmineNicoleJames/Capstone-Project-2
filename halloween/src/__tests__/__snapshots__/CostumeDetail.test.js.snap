// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h3
        class="result-text"
      >
         Watch this tutorial to learn how to transform yourself into 
      </h3>
      <p
        class="result-text"
      >
        FAKE
      </p>
      <iframe
        allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
        frameborder="0"
        height="315"
        width="560"
      />
      <p
        class="title"
      />
    </div>
  </div>
</DocumentFragment>
`;
